subsys/usb/device_next/class/usbd_msc_scsi.c

```diff
@@ -0,0 +1,833 @@
+#define INQUIRY_VERSION_SPC_5	0x07
+
+/* Claim conformance to SPC-2 because this allows us to implement less commands
+ * and do not care about multiple reserved bits that became actual options
+ * later on. DO NOT change unless you make sure that all mandatory commands are
+ * implemented and all options (e.g. vpd pages) that are mandatory at given
+ * version are also implemented.
+ */
+#define CLAIMED_CONFORMANCE_VERSION INQUIRY_VERSION_SPC_2
+
+#define T10_VENDOR_LENGTH	8
+#define T10_PRODUCT_LENGTH	16
+#define T10_REVISION_LENGTH	4
+
+/* Optional, however Windows insists on reading Unit Serial Number.
+ * There doesn't seem to be requirement on minimum product serial number length,
+ * however when the number is not available the device shall return ASCII spaces
+ * in the field.
+ */
+#define UNIT_SERIAL_NUMBER "  "
+
+/* Every SCSI command has to abide to the general handling rules. Use macros
+ * to allow generating boilerplate handling code.
+ */
+#define SCSI_CMD_STRUCT(opcode) struct scsi_##opcode##_cmd
+#define SCSI_CMD_HANDLER(opcode)					\
+static int scsi_##opcode(struct scsi_ctx *ctx,				\
+			 struct scsi_##opcode##_cmd *cmd,		\
+			 uint8_t data_in_buf[CONFIG_USBD_MSC_SCSI_BUFFER_SIZE])
+
+/* SAM-6 5.2 Command descriptor block (CDB)
+ * Table 43 – CONTROL byte
+ */
+#define GET_CONTROL_NACA(cmd)		(cmd->control & BIT(2))
+
+/* SPC-5 4.3.3 Variable type data field requirements
+ * Table 25 — Code set enumeration
+ */
+enum code_set {
+	CODE_SET_BINARY = 0x1,
+	CODE_SET_ASCII = 0x2,
+	CODE_SET_UTF8 = 0x3,
+};
+
+/* SPC-5 F.3.1 Operation codes Table F.2 — Operation codes */
+enum scsi_opcode {
+	TEST_UNIT_READY = 0x00,
+	REQUEST_SENSE = 0x03,
+	INQUIRY = 0x12,
+	MODE_SENSE_6 = 0x1A,
+	START_STOP_UNIT = 0x1B,
+	PREVENT_ALLOW_MEDIUM_REMOVAL = 0x1E,
+	READ_FORMAT_CAPACITIES = 0x23,
+	READ_CAPACITY_10 = 0x25,
+	READ_10 = 0x28,
+	WRITE_10 = 0x2A,
+};
+
+SCSI_CMD_STRUCT(TEST_UNIT_READY) {
+	uint8_t opcode;
+	uint32_t reserved;
+	uint8_t control;
+} __packed;
+
+/* DESC bit was reserved in SPC-2 and is optional since SPC-3 */
+#define GET_REQUEST_SENSE_DESC(cmd)	(cmd->desc & BIT(0))
+
+SCSI_CMD_STRUCT(REQUEST_SENSE) {
+	uint8_t opcode;
+	uint8_t desc;
+	uint8_t reserved2;
+	uint8_t reserved3;
+	uint8_t allocation_length;
+	uint8_t control;
+} __packed;
+
+#define SENSE_VALID			BIT(7)
+#define SENSE_CODE_CURRENT_ERRORS	0x70
+#define SENSE_CODE_DEFERRED_ERRORS	0x71
+
+#define SENSE_FILEMARK			BIT(7)
+#define SENSE_EOM			BIT(6)
+#define SENSE_ILI			BIT(5)
+#define SENSE_KEY_MASK			BIT_MASK(4)
+
+#define SENSE_KEY_SPECIFIC_VALID	BIT(16)
+#define SENSE_KEY_SPECIFIC_MASK		BIT_MASK(15)
+
+struct scsi_request_sense_response {
+	uint8_t valid_code;
+	uint8_t obsolete;
+	uint8_t filemark_eom_ili_sense_key;
+	uint32_t information;
+	uint8_t additional_sense_length;
+	uint32_t command_specific_information;
+	uint16_t additional_sense_with_qualifier;
+	uint8_t field_replaceable_unit_code;
+	uint16_t sense_key_specific;
+	uint8_t additional_sense_byte;
```

Comment:
Would you consider shortening those? In Bluetooth we do it to avoid neverending variable names like those

Reply:
There is problem with coming up with names for things that are defined in SCSI documents (here SPC-5, Table 48 Fixed format sense data). I mostly followed the names from the standard (`additional_sense_with_qualifier` is actually merged `additional sense code` and `additional sense code qualifier` because the two are actually defined in one table). I don't think it matters much because these defines are used in only one place (here: when composing the response).

One alternative to avoid structures like this one is to just write the fields in order like it is done in `fill_vpd_page()` where the structures would get pretty much insane.

Actually there is a bug in here, `uint8_t sksv;` should be added before `uint16_t sense_key_specific;`. From the main table it looks like it is 24-bit field but really it is used as 8-bit bitfield (with lower 7 bits changing meaning based on sense key) followed by 16-bit value.

Reply:
Why is there a problem with shortening names that are defined in a spec? We do it all the time in Bluetooth. You could do something like:

```
uint8_t valid_code;
	uint8_t obsolete;
	uint8_t filemark_eom_ili_sense_key;
	uint32_t info;
	uint8_t add_sense_len;
	uint32_t cmd_specific_infor;
	uint16_t add_sense_w_qual;
	uint8_t field_replac_unit_code;
	uint16_t sense_key_spec;
	uint8_t add_sense_byte;
```

Reply:
What is the actual value behind shorter names? I cannot really see the benefit. However, for the shortened names, the prime example is the old USB MSC implementation where it is hard to tell where each field comes from - because it not only mixes SCSI (big-endian) and USB together (little-endian), but also comes up with shortened names.

Reply:
The value is just to avoid having strings that are unwieldy, but I won't insist beyond the suggestion, given the limited benefit as you point out.